### PR TITLE
Canvas text and macOS font fixes

### LINKED ISSUE #24622: Intermittent mac timeout due to missing font URLs (No URL for Core Text font!)
This affects lots of different tests, so I'm not adding any tests to the summary.
```
 ▶ TIMEOUT [expected FAIL] /css/css-fonts/font-variant-numeric-08.html
  │ 
  │ 
  │ 
  │ 
  │ No URL for Core Text font! (thread FontCacheThread, at src/libcore/option.rs:1185)
  │ stack backtrace:
  │    0: backtrace::backtrace::trace
  │    1: backtrace::capture::Backtrace::new
  │    2: servo::main::{{closure}}
  │    3: std::panicking::rust_panic_with_hook
  │    4: std::panicking::continue_panic_fmt
  │    5: rust_begin_unwind
  │    6: core::panicking::panic_fmt
  │    7: core::option::expect_failed
  │    8: gfx::platform::macos::font_template::FontTemplateData::bytes
  │    9: gfx::font_cache_thread::FontCache::run
  │   10: std::sys_common::backtrace::__rust_begin_short_backtrace
  │   11: std::panicking::try::do_call
  │   12: __rust_maybe_catch_panic
  │   13: core::ops::function::FnOnce::call_once{{vtable.shim}}
  │   14: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  │   15: std::sys::unix::thread::Thread::new::thread_start
  │   16: _pthread_body
  │   17: _pthread_start
```

### LINKED ISSUE #27476: Fail to get font handle in canvas fill_text
While trying to access [this page from a Taiwanese magazine publisher](https://web.cw.com.tw/corporate-political-money-2020/index.html) for showing some data with D3, I can't see the page rendered correctly and will see following panic message.

<details>

<summary>Open to see full backtrace</summary>

```
→ ./mach run https://web.cw.com.tw/corporate-political-money-2020/index.html
[2020-08-01T10:32:52Z ERROR canvas::canvas_data] error getting font handle for style Font { font_family: FontFamily { families: FontFamilyList([FamilyName(FamilyName { name: Atom('Noto Sans TC' type=dynamic), syntax: Quoted })]), is_system_font: false }, font_style: Normal, font_variant_caps: Normal, font_weight: FontWeight(400.0), font_size: FontSize { size: NonNegative(20.0 px), keyword_info: KeywordInfo { kw: None, factor: 1.0, offset: 0.0 px } }, font_stretch: FontStretch(NonNegative(Percentage(1.0))), hash: 930802703126591095 }: no font found
called `Option::unwrap()` on a `None` value (thread CanvasThread, at /Users/cybai/.cargo/git/checkouts/raqote-9e18d1f68843f9af/ff5d2d1/src/draw_target.rs:755)
   0: backtrace::backtrace::libunwind::trace
             at /Users/cybai/.cargo/git/checkouts/backtrace-rs-fb8cc3ab7e3059b0/5366f72/src/backtrace/libunwind.rs:96
      backtrace::backtrace::trace_unsynchronized
             at /Users/cybai/.cargo/git/checkouts/backtrace-rs-fb8cc3ab7e3059b0/5366f72/src/backtrace/mod.rs:66
   1: <servo::backtrace::Print as core::fmt::Debug>::fmt
             at /Volumes/Transcend/codespace/mozilla/servo/ports/winit/backtrace.rs:49
   2: core::fmt::write
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libcore/fmt/mod.rs:1117
   3: std::io::Write::write_fmt
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/io/mod.rs:1508
   4: servo::backtrace::print
             at /Volumes/Transcend/codespace/mozilla/servo/ports/winit/backtrace.rs:17
   5: servo::main::{{closure}}
             at /Volumes/Transcend/codespace/mozilla/servo/ports/winit/main2.rs:166
   6: std::panicking::rust_panic_with_hook
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/panicking.rs:530
   7: rust_begin_unwind
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/panicking.rs:437
   8: core::panicking::panic_fmt
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libcore/panicking.rs:85
   9: core::panicking::panic
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libcore/panicking.rs:50
  10: core::option::Option<T>::unwrap
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libcore/macros/mod.rs:10
  11: raqote::draw_target::DrawTarget::draw_text
             at /Users/cybai/.cargo/git/checkouts/raqote-9e18d1f68843f9af/ff5d2d1/src/draw_target.rs:755
  12: canvas::raqote_backend::<impl canvas::canvas_data::GenericDrawTarget for raqote::draw_target::DrawTarget>::fill_text
             at /Volumes/Transcend/codespace/mozilla/servo/components/canvas/raqote_backend.rs:530
  13: canvas::canvas_data::CanvasData::fill_text
             at /Volumes/Transcend/codespace/mozilla/servo/components/canvas/canvas_data.rs:570
  14: canvas::canvas_paint_thread::CanvasPaintThread::process_canvas_2d_message
             at /Volumes/Transcend/codespace/mozilla/servo/components/canvas/canvas_paint_thread.rs:148
  15: canvas::canvas_paint_thread::CanvasPaintThread::start::{{closure}}
             at /Volumes/Transcend/codespace/mozilla/servo/components/canvas/canvas_paint_thread.rs:73
  16: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/sys_common/backtrace.rs:130
  17: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/thread/mod.rs:475
  18: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/panic.rs:318
  19: std::panicking::try::do_call
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/panicking.rs:348
  20: ___rust_try
  21: std::panicking::try
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/panicking.rs:325
  22: std::panic::catch_unwind
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/panic.rs:394
  23: std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/thread/mod.rs:474
  24: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libcore/ops/function.rs:233
  25: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/liballoc/boxed.rs:1074
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/liballoc/boxed.rs:1074
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/6c8927b0cf80ceee19386026cf9d7fd4fd9d486f/src/libstd/sys/unix/thread.rs:87
  26: __pthread_body
  27: __pthread_start
[2020-08-01T10:34:15Z ERROR servo] called `Option::unwrap()` on a `None` value
```

</details>

| Firefox | Servo |
| --- | --- |
| image | image |

Built with latest main branch (6499367fe2) on macOS 10.14.6

### PR DESCRIPTION
**Pull request by jdm**
These changes work around a source of panics by skipping glyphs that don't exist when drawing text to the canvas. Ideally we would have font fallback and/or tofu characters, but that's a larger project than I wanted to get into right now.

Relatedly, while investigating the non-panic errors observed in #27476, I realized that the underlying source of #24622 could be dealt with because we shouldn't actually need to open the CTFont by a file path to get the bytes, since we already have bytes available that we use to create the CTFont.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24622 and fix #27476 and fix the panic observed in #20445.
- [x] There are tests for these changes

### CODE REVIEW

File: components/canvas/raqote_backend.rs [hunk 1]
```diff
@@ -527,11 +527,33 @@ impl GenericDrawTarget for raqote::DrawTarget {
...
+            ids.push(id);
+            positions.push(Point2D::new(start.x(), start.y()));
+            let advance = match font.advance(id) {
+                Ok(advance) => advance,
+                Err(e) => {
+                    warn!("Skipping glyph {} with missing advance: {:?}", c, e);
+                    continue;
+                },
+            };
+            start += advance * point_size / 24. / 96.;
```
Comment (pcwalton) [inline components/canvas/raqote_backend.rs:550]: Where do these values come from? Are you sure this isn't the `units_per_em` for some font?

Comment (jdm) [reply components/canvas/raqote_backend.rs:550]: This is copied directly from the code we used to call in https://github.com/jrmuizel/raqote/blob/490bff22610a130ad75eb797b6bd6e3e361bbde5/src/draw_target.rs#L742-L761.

**Pull request merged**